Lib/test/test_asyncio/test_tasks.py

```diff
@@ -3681,15 +3682,21 @@ def task_factory(loop, coro):
         self.assertEqual(context['exception'], exc_context.exception)
 
     def test_run_coroutine_threadsafe_and_cancel(self):
-        async def target():
-            # self.loop.run_in_executor(None, _in_another_thread)
-            thread_future = asyncio.run_coroutine_threadsafe(self.add(1, 2), self.loop)
-            await asyncio.sleep(0)
+        target_started = Event()
+        async def _target():
+            target_started.set()
+            await asyncio.sleep(0.1)
+            return 1
 
+        def _in_thread():
+            thread_future = asyncio.run_coroutine_threadsafe(_target(), self.loop)
+            # wait target started then cancel
+            target_started.wait()
             thread_future.cancel()
-            await asyncio.sleep(0)
+        _ = self.loop.run_in_executor(None, _in_thread)
 
-        self.loop.run_until_complete(target())
+        # start main loop to do things
+        self.loop.run_until_complete(asyncio.sleep(0.05))
```

Comment:
~I think I have a way to fix the test and stop relying on unreliable timeouts:~


```suggestion
        future = self.loop.run_in_executor(None, _in_thread)

        # start main loop to do things
        self.loop.run_until_complete(future)
        self.loop.run_until_complete(asyncio.sleep(0))
```

Reply:
Unfortunately, this way the test also passes *without* your fix in futures.py, so no, this is not the fix we're looking for... :-(

Reply:
Hi @gvanrossum  , the issue in [gh-105836](https://github.com/python/cpython/issues/105836) only occurs while run asyncio.run_coroutine_threadsafe in the loop/main thread. The example in that issue is this case because it use Thread.run() instead of Thread.start(). Maybe I should change the test case back to run in main thread?

Reply:
Oh you're right! My bad.